### PR TITLE
Use StylistIdentifier instead of just strings

### PR DESCRIPTION
**Pull request by kerrmarin**
Allows us to create WithStylistIdentifier with a set of identifiers as well as just strings

### CODE REVIEW

File: StylableSwiftUI/Classes/WithStyleIdentifier.swift [hunk 1]
```diff
@@ -60,31 +65,65 @@ public extension WithStylistIdentifier {
...
                 @ViewBuilder contents: @escaping (StylistIdentifier, StylistIdentifier, StylistIdentifier) -> Content) {
-        self.tokens = [ token1, token2, token3 ]
+        self.tokens = [ StylistIdentifier(token1), StylistIdentifier(token2), StylistIdentifier(token3) ]
         self.contents = { contents($0[0], $0[1], $0[2]) }
     }
 
     init(tokens token1: String, _ token2: String, _ token3: String, _ token4: String,
                 @ViewBuilder contents: @escaping (StylistIdentifier, StylistIdentifier, StylistIdentifier, StylistIdentifier) -> Content) {
-        self.tokens = [ token1, token2, token3, token4 ]
+        self.tokens = [ StylistIdentifier(token1), StylistIdentifier(token2), StylistIdentifier(token3), StylistIdentifier(token4) ]
```
Comment (deanWombourne) [inline StylableSwiftUI/Classes/WithStyleIdentifier.swift]: Could these four init methods just call out to the other ones i.e.

```swift
self.init(tokens: tokens.map(StylistIdentifier.init), contents: contents)
```

I can't tell if that complication is worse than the duplication :) Or if it's so small it doesn't matter!

Comment (deanWombourne) [reply StylableSwiftUI/Classes/WithStyleIdentifier.swift]: oh, no, no they can't - damn variadic arguments!

File: StylableSwiftUI/Classes/WithStyleIdentifier.swift [hunk 2]
```diff
@@ -37,19 +37,24 @@ public struct WithStylistIdentifier<Content: View>: View {
 
     @Environment(\.currentStylableGroup) private var currentStylableGroup
 
-    private let tokens: [String]
+    private let tokens: [StylistIdentifier]
```
Comment (deanWombourne) [inline StylableSwiftUI/Classes/WithStyleIdentifier.swift]: This parameter doesn't make sense being called "token" anymore :)

**Review by deanWombourne — CHANGES_REQUESTED** [review]
Any chance of some tests ;)

**Comment by kerrmarin** [comment]
@deanWombourne I added a view with the new option of using the stylist identifiers but I'm not sure there are any tests that could be written to test the _actual_ view?

File: StylableSwiftUI/Classes/WithStyleIdentifier.swift (hunk 2 above)
Comment (kerrmarin) [reply StylableSwiftUI/Classes/WithStyleIdentifier.swift]: Done

**Comment by deanWombourne** [comment]
Hmm, that's a fair point. Would be nice to have a unit test which could take a WithStylistIdentifier created with strings and check they are the correct identifiers (i.e. they have been converted and then appended onto the existing stylist group) but I don't know enough about SwiftUI vs. testing to know how to do that.

Oh, that's just made me think of an edge case :) What would you expect this to do?

```swift
WithStylistIdentifier(token: "hi/there") { identifier in }
```

Would you expect `identifier` to be an identifier with the token "hi/there", or would you expect identifier to be an identifier with a single path component "hi" and the token "there" ?

If it's the latter we should probably change the parameter name from `token` to `string`. If the former, we might need to use a more specific init method on StylistIdentifier?

**Comment by kerrmarin** [comment]
It's the latter, that's actually the use case I'm trying to solve -- using the original method with a string with forward slashes didn't convert them to a path :)

**Review by deanWombourne — APPROVED** [review]

**Comment by deanWombourne** [comment]
I _think_ that StylistIdentifier is string convertible, so now you can also just do `WithStylistIdentifier(strings: "a/full/path", "another/one") { ... }`?

**Comment by kerrmarin** [comment]
yup

**Pull request merged**